### ISSUE TITLE
Add support for Nix, Guix, et al.

### ISSUE THREAD

**Issue by thblt** [emacscollective/borg#10]
org-mode (and maybe other packages) is built with a Makefile.  Maybe `borg-build` could do a `make` if a Makefile is present?

**Comment by tarsius** [comment]
Just because a repository contains a `Makefile`, that does not necessarily mean that we *have* to use it. `magit` for example comes with a `Makefile`, but it's not necessary to use it.

And when `make` actually has to be used, such as in the case of `org` and `mu4e`, then someone has to figure out what targets to run and store that information somewhere. We can't just blindly run `make` and hope for the best. Also note that `borg` might have to be told about unusual `load-path`s.

`borg` supports custom build-steps and `load-path`s, but has to be told about them. This is done in `~/.emacs.d/.gitmodules` as described in https://emacsmirror.net/manual/borg/Git-variables.html.

The plan is to eventually provide ready-to-use configurations for some popular packages in the form of "drone branches". Such a branch would assimilate a package (and its dependencies if necessary) and set the necessary variables in `~/.emacs.d/.gitmodules`. Such a package could then be assimilated by merging the drone branch instead of using `borg-assimilate`.

I might create such drone branches for the mentioned packages in a little while, but for the time being look at the documentation (these two packages are used as examples). I also intend to improve the documentation on how to deal with non-trivial packages and the use of drone branches.

**Comment by thblt** [comment]
> borg supports custom build-steps and load-paths, but has to be told about them. This is done in ~/.emacs.d/.gitmodules as described in https://emacsmirror.net/manual/borg/Git-variables.html.

Thanks, I've skipped that part.

> I might create such drone branches for the mentioned packages in a little while.

Would you be interested in adding Nix support to the build process? Something like `submodule.DRONE.build-nix-shell` (for a list of Nix packages) and `submodule.DRONE.build-nix-shell-file` (for a shell file). I cannot promise I'll do it fast, but I'd be happy to help on that.

Thanks a lot for your answer, I guess this can be closed.

**Comment by tarsius** [comment]
While I am interested in Nix (and Guix), I don't really know what that means.

Wouldn't `submodule.DRONE.build-nix-shell = ARGS` be equivalent to `submodule.DRONE.build-step = nix-shell ARGS`? If so, then I don't think any special support should be added.

Even if not, I think this should - at least initially - be a patch and not be merged into `master`.

**Comment by thblt** [comment]
> Wouldn't submodule.DRONE.build-nix-shell = ARGS be equivalent to submodule.DRONE.build-step = nix-shell ARGS? If so, then I don't think any special support should be added.

Yes, it would be equivalent, but it would *require* Nix (or Guix, which I've never tested).   On the other hand, `submodule.DRONE.build-step = make` would be unusable on NixOS unless the package has no external library dependency at all, since NixOS doesn't make *any* library available on the default environment.  (I'm thinking of packages with compiled helper programs, like Irony or PDF Tools.)

My idea was to provide a configuration option to start a Nix shell and execute all the build steps inside this shell if Nix is available/somehow enabled.  This would make the "drone branches" you were referring to more portable.  

A probably better option would be to provide a generic framework for Nix-Shell-like wrappers, where the actual implementations would be provided as third-party packages, and default to the current behavior, without handling the special cases within Borg.   I'm thinking of something like:

``` elisp
(defun borg-run-build-steps-shell (drone)
    "Execute build steps for DRONE, in order, through `shell-command'."
    ...)

(defvar borg-run-build-steps nil
    "The command to run the build-steps for a drone.  The value should be a function taking one argument, the drone name, and returning non-nil if all steps ran successfully."
    borg-run-build-steps-shell
```

**Comment by tarsius** [comment]
Yes, please implement that. (I would do it in two commits a nix specific one (just so that we see what that would look like) and then one that generalizes the approach. Later we can still squash the two.)

**Comment by tarsius** [comment]
I have implemented this in cefde4207192df158909537c23be818e4a9d9d37. Please check it out.

Thanks for the suggestion and working on this!
